[RNMobile] Fixed position of `FloatingToolbar` on the bottom of the screen

jbinda opened the pull request:
## Description
<!-- Please describe what you have changed or added -->
This PR changes behaviour of `FloatingToolbar` to align with guideline from [this issue](https://github.com/wordpress-mobile/gutenberg-mobile/issues/1414#issuecomment-603972811). It will be positioning absolutely at the bottom of the screen.

Please also refer to:
[Related gutenberg-mobile PR](https://github.com/wordpress-mobile/gutenberg-mobile/pull/2118)
[Related gutenberg-mobile issue](https://github.com/wordpress-mobile/gutenberg-mobile/issues/1414#issuecomment-603972811)

Testing branch on WP iOS and WP Android:
WordPress-iOS:  [here](https://github.com/wordpress-mobile/WordPress-iOS/pull/13927#issuecomment-615245960)
WordPress-Android: [here](https://github.com/wordpress-mobile/WordPress-Android/pull/11692#issuecomment-615247791)


## How has this been tested?
<!-- Please describe in detail how you tested your changes. -->
<!-- Include details of your testing environment, tests ran to see how -->
<!-- your change affects other areas of the code, etc. -->


1. Add below code in `initial-html`

<details>

```
/**
 * @format
 * @flow
 */

export default `
<!-- wp:media-text {"mediaId":1,"mediaType":"image"} -->
<div class="wp-block-media-text is-stacked-on-mobile"><figure class="wp-block-media-text__media"><img src="https://cldup.com/cXyG__fTLN.jpg" class="wp-image-1"/></figure><div class="wp-block-media-text__content"><!-- wp:paragraph {"className":"has-large-font-size"} -->
<p class="has-large-font-size"></p>
<!-- /wp:paragraph --></div></div>
<!-- /wp:media-text -->

<!-- wp:image -->
<figure class="wp-block-image"><img src="https://cldup.com/cXyG__fTLN.jpg" alt=""/></figure>
<!-- /wp:image -->

<!-- wp:group -->
<div class="wp-block-group"><div class="wp-block-group__inner-container"></div></div>
<!-- /wp:group -->

<!-- wp:group -->
<div class="wp-block-group"><div class="wp-block-group__inner-container"><!-- wp:image {"id":1,"sizeSlug":"large"} -->
<figure class="wp-block-image size-large"><img src="https://cldup.com/cXyG__fTLN.jpg" alt="" class="wp-image-1"/></figure>
<!-- /wp:image -->

<!-- wp:heading -->
<h2>Test heading</h2>
<!-- /wp:heading --></div></div>
<!-- /wp:group -->
`;
```

</details>

2. Feel free to add more nested structure
3. Check if each time you select nested block FloatingToolbar is visible on bottom of the screen



## Screenshots <!-- if applicable -->
<img src="https://user-images.githubusercontent.com/21242757/78678005-af7a6000-78e8-11ea-8185-1c979bf11f33.gif" width="300" />


**LTR vs RTL mode**

<img width="250" src="https://user-images.githubusercontent.com/21242757/78877062-8ffe4700-7a50-11ea-8583-21e75a58d34f.png"><img width="265" src="https://user-images.githubusercontent.com/21242757/78877053-8d035680-7a50-11ea-90d7-1bad262d5773.png">

## Types of changes
<!-- What types of changes does your code introduce?  -->
<!-- Bug fix (non-breaking change which fixes an issue) -->
<!-- New feature (non-breaking change which adds functionality) -->
<!-- Breaking change (fix or feature that would cause existing functionality to not work as expected) -->

Refactor - change the FloatingToolbar placement and behaviour

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/javascript/ -->
- [x] My code follows the accessibility standards. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/accessibility-coding-standards/ -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/javascript/ -->
- [x] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR. <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/master/docs/contributors/native-mobile.md -->

jbinda committed:
move FloatingToolbar to slot-fill architecture

jbinda committed:
Merge remote-tracking branch 'origin/master' into rnmobile/feat/fixed-floating-toolbar-position

jbinda commented on packages/block-editor/src/components/block-list/block.native.js:
```diff
@@ -113,21 +112,19 @@ class BlockListBlock extends Component {
-					) }
+					<FloatingToolbar clientId={ clientId }>
+						{ hasParent && (
+							<Toolbar passedStyle={ styles.toolbar }>
+								<ToolbarButton
+									title={ __( 'Navigate Up' ) }
+									onClick={ () => onSelect( parentId ) }
+									icon={ NavigateUpSVG }
+								/>
+								<View style={ styles.pipe } />
+							</Toolbar>
+						) }
+						<Breadcrumbs clientId={ clientId } />
```
with slot-fill architecture on `FloatingToolbar` we can also move all the content to FloatingToolbar file and use it here as below:

`<FloatingToolbar clientId={ clientId } />`

Does it make sense or we would like to keep current PR version?

jbinda replied:
@dratwas I refactor to above according to our conversation

jbinda commented on packages/block-editor/src/components/block-list/index.native.js:
```diff
@@ -121,6 +122,7 @@ export class BlockList extends Component {
 				style={ containerStyle }
 				onAccessibilityEscape={ clearSelectedBlock }
 			>
+				{ isRootList && <FloatingToolbar.Slot /> }
```
Rendering slot only for RootList should eliminate render FloatingToolbar for each nested level

dratwas replied:
Hmm, maybe we could move it somewhere else. i.e `packages/edit-post/src/components/layout/index.native.js`. Do you think it is possible?

jbinda replied:
As discussed on call - moved it to layout component

jbinda commented on packages/block-editor/src/components/block-mobile-floating-toolbar/styles.native.scss:
```diff
@@ -12,7 +13,7 @@ $floating-toolbar-height: 44;
 	justify-content: center;
 	align-self: center;
-	margin-bottom: 8px;
+	bottom: 51;
```
This is temporary to set proper space of 8px between bottom toolbar and FloatingToolbar. Will refactor soon

jbinda replied:
fixed!

jbinda committed:
handle proper FloatingToolbar positioning iOS vs Android

move FloatingToolbar fill from BlockListBLock component

cameronvoell commented:
Found a new crash on this PR tapping the toolbar button to split a Group block into separate blocks.
Steps: 
1. Add a group block with one or more blocks
2. Tap Arrow to select entire Group block
3. Tap toolbar button to split group block into separate blocks
4. Crash, see screen shot below

![image](https://user-images.githubusercontent.com/1103838/78827284-b3b18680-7997-11ea-9813-9f5e57fe78d7.png)

jbinda committed:
Merge remote-tracking branch 'origin/master' into rnmobile/feat/fixed-floating-toolbar-position

jbinda commented:
thanks for reporting ! Will check that today

jbinda committed:
fix crash after ungroup or delete block

jbinda committed:
support RTL mode by rotate the icons

jbinda commented:
I also noticed that it crash after try to remove any block. Already fixed that. 

The cause was that after ungroup (or delete) block breadcrumbs can not read block icon.

I have also add suport for RTL according to [this comment](https://github.com/wordpress-mobile/gutenberg-mobile/issues/2133#issuecomment-610998527)

jbinda committed:
Merge remote-tracking branch 'origin/master' into rnmobile/feat/fixed-floating-toolbar-position

dratwas commented on packages/block-editor/src/components/block-list/breadcrumb.native.js:
```diff
@@ -91,13 +95,14 @@ export default compose( [
 			blockIcon,
 			rootClientId,
 			rootBlockIcon,
+			isRTL: I18nManager.isRTL,
```
Why you added it here instead of using it directly? You could even move this logic to the `SubdirectorSVG`. WDYT?

jbinda replied:
I refactored to use it directly and keep login here according to what we discussed

dratwas commented on packages/block-editor/src/components/block-mobile-floating-toolbar/index.native.js:
```diff
@@ -0,0 +1,83 @@
+/**
```
This is a mobile-specific component. Please move it to the `packages/components/mobile/`

jbinda replied:
moved

dratwas commented on packages/block-editor/src/components/block-mobile-floating-toolbar/nav-up-icon.js:
```diff
@@ -3,7 +3,7 @@
 import { SVG, Path } from '@wordpress/components';
 
-const NavigateUp = (
+const NavigateUp = ( { isRTL } ) => (
```
Since it is only used for mobile I suggest to move the import of  I18nManager here and remove it from the floating toolbar

jbinda replied:
I keep it according to our discussion before

dratwas commented on packages/block-editor/src/components/block-mobile-floating-toolbar/styles.native.scss:
```diff
@@ -0,0 +1,43 @@
+	border-left-width: 0;
+	margin-right: auto;
+	padding-left: 2px;
+	height: 44px;
```
```suggestion
	height: $mobile-floating-toolbar-height;
```

jbinda replied:
done.

jbinda committed:
changes after code review

pinarol commented:
@jbinda Could you test with Keyboard open? The FloatingToolbar will probably cover the caret on a Paragraph block.

On Android, the autoscroll is system managed, we don't do a custom thing for it so it can be a problem. 

On iOS, we are doing it with custom implementation but it can still be a problem because we haven't tried setting the extra scroll amount to a dynamic value so far. We now need to add the height of the FloatingToolbar to the extra scroll amount when it is visible.

jbinda committed:
move FloatingToolbar to packages/components/src/mobile

jbinda commented:
@pinarol I have do some testing and below is the results:

When you trying to select multiline text based block you can noticed that after scroll FloatingToolbar will cover the caret in text field - see below GIFs

<img src="https://user-images.githubusercontent.com/21242757/79358009-00213700-7f41-11ea-9884-8e3d8d7aac3a.gif" width="250" />

<img src="https://user-images.githubusercontent.com/21242757/79358088-18915180-7f41-11ea-96d5-9709a84acc49.gif" width="250" />

On iOS we can solve that by adding FloatingToolbar height when it's visible to `inputAccessoryViewHeight`prop in `KeyboardAwareFlatList` as @pinarol suggested. On Android we currently do not have solution for that.


Alternative solution works for both is putting the `FLoatingToolbar` to the top of the screen. It solves mentioned problems. See below GIFs

<img src="https://user-images.githubusercontent.com/21242757/79357338-2db9b080-7f40-11ea-9a98-fc0ef8d09dbc.gif" width="250" />   

<img src="https://user-images.githubusercontent.com/21242757/79357365-34e0be80-7f40-11ea-950b-f179b0c1eafd.gif" width="250" />

@iamthomasbishop what do you think ?

jbinda commented:
Another think is that I have noticed different behaviour depending on where we render `FloatingToolbar.SLot`

Initially I have render`<FloatingToolbar.SLot />` in BlockListBlock like this `{isRootList && <FloatingToolbar.Slot /> }` but then we will need to adjust position on iOS after keyboard opens. Without that it will be covered (also the position of additional toolbar is wrong)

<img src ="https://user-images.githubusercontent.com/21242757/79358695-d9173500-7f41-11ea-83d5-7e8d7ab891e8.gif" width="250" />

With latest discussion with @dratwas we conclude that we can move `<FloatingToolbar.Slot />` to layout component. At first glance it works fine (something stays in cache thats why I didn't spot that earlier) however after checking this with keyboard I also noticed that on Android it stops to be focusable at all.

jbinda commented on packages/edit-post/src/components/layout/index.native.js:
```diff
@@ -151,6 +154,7 @@ class Layout extends Component {
 						) }
 						<Header />
 						<BottomSheetSettings />
+						<FloatingToolbar.Slot />
```
It seems moving `FloatingToolbar.Slot` here make it unfocusable at all on Android (see [this comment](https://github.com/WordPress/gutenberg/pull/21446#issuecomment-614136356))

jbinda committed:
fix lint

jbinda committed:
Merge remote-tracking branch 'origin/master' into rnmobile/feat/fixed-floating-toolbar-position

pinarol commented:
@iamthomasbishop  unfortunately we don’t have a customizable scroll mechanism on Android. Android just adjusts itself to the viewport, what I mean by viewport is this red frame:

<img width="250" alt="Screen Shot 2020-04-16 at 12 29 23" src="https://user-images.githubusercontent.com/5032900/79447123-0f0ef480-7fe8-11ea-993d-48889813ac9e.png">

I tried a simple mock to demonstrate how it’d look if we added the height of floating toolbar(assume it is the yellow view) to the equation:

<img width="250" alt="Screen Shot 2020-04-16 at 13 16 07" src="https://user-images.githubusercontent.com/5032900/79447140-17672f80-7fe8-11ea-9a61-d0796f298429.png">

you see how it just clips the viewport height?
so actually the whole mechanism just works by clipping the viewport, there’s no actual “extra scrolling” going on that we can control.

on iOS we do have that extra scrolling mechanism because we had to implement the whole autoscroll mechanism from scratch. We had no choice since on iOS we don’t have a system supported keyboard avoiding mechanism(at least on the components react-native allows us to use). it took months to get it working the way it works now.

on Android it is all system managed, we just clip the viewport and it automatically adjusts the scroll position according to where the caret is

So if we want to keep the FloatingToolbar at the bottom, we can consider dedicating the whole rectangle area to it by changing the design. (forget about the yellow view for a second, imagine the whole red frame is the floating toolbar, except that it is not floating any more :( )

<img width="250" alt="Screen Shot 2020-04-16 at 13 17 44" src="https://user-images.githubusercontent.com/5032900/79447244-3ebdfc80-7fe8-11ea-8273-8c94744e2ac7.png">

pinarol commented:
> @pinarol Ok, that all makes sense. I think at this point, our options are to either consider a new design that would take up the full-width in an area at top of the keyboard or to anchor the UI at the top of the viewport. Is it possible to keep iOS bottom-positioned and Android top (because of the restrictions you mentioned above)? I don't think this is ideal, but it isn't the worst option. It could also allow us to test out both variations and see which performs better.

@iamthomasbishop You mean as a permanent solution? I am all cool with having a few releases with such a difference to see which one we like. But, at the end of this phase I'd really prefer if we decide on a single UI. Having different UI per platform doubles our maintenance efforts. (BTW, our maintenance efforts our are already quite high(~50%) so the more we increase it the less time we'll have for new features)

Do you want us to generate new builds for each platform so you can test both options side by side? iOS bottom-positioned and Android top. Maybe one option will stand out quickly when you try.

iamthomasbishop commented:
> You mean as a permanent solution?

No, a temporary one.

> Do you want us to generate new builds for each platform so you can test both options side by side? iOS bottom-positioned and Android top. Maybe one option will stand out quickly when you try.

Precisely. I'm pretty certain bottom-positioned is going to win out, but until we solve the positioning challenge, this is workable.

jbinda committed:
render FloatingToolbar top on Android bottom on iOS

jbinda committed:
apply additional offset to autoscroll on iOS

jbinda committed:
Merge remote-tracking branch 'origin/master' into rnmobile/feat/fixed-floating-toolbar-position

jbinda committed:
remove slot-fill architecture from FloatingToolbar

jbinda commented:
@pinarol @iamthomasbishop 

I have update the FloatingToolbar according to discussion and now it works like below:

<img src="https://user-images.githubusercontent.com/21242757/79554141-f103cb80-809d-11ea-89e7-6b1cbbce4590.gif" width="300" />

<img src="https://user-images.githubusercontent.com/21242757/79554403-4a6bfa80-809e-11ea-93f8-c832f683a12c.gif" width="300" />

jbinda committed:
Merge remote-tracking branch 'origin/master' into rnmobile/feat/fixed-floating-toolbar-position

jbinda committed:
clean code

jbinda committed:
Merge remote-tracking branch 'origin/master' into rnmobile/feat/fixed-floating-toolbar-position

pinarol commented:
> No, a temporary one.

 Cool 👌 

> but until we solve the positioning challenge

@iamthomasbishop Just for sake of clarity, does this mean until we come up with a new design that would take up the full-width in an area at top of the keyboard ?

jbinda committed:
move floating toolbar to block-edit package

pinarol commented:
@jbinda Thank you for the updates. So, this is ready for another set of code reviews right?

jbinda commented:
I will push updates in a minutes because I was dealing with the CI fails.

It's because how I import/export Breadcrumbs in FloatingToolbar. After refactor CI should pass and I will be able to generate bundles

jbinda commented on packages/block-editor/src/components/block-list/index.native.js:
```diff
@@ -132,7 +138,12 @@ export class BlockList extends Component {
 					}
-					inputAccessoryViewHeight={ headerToolbar.height }
+					inputAccessoryViewHeight={
+						headerToolbar.height +
+						( isFloatingToolbarVisible
+							? floatingToolbar.height
+							: 0 )
+					}
```
This will add additional space for FloatingToolbar when autoscroll to text base blocks

jbinda committed:
fix lint

jbinda commented:
@pinarol @iamthomasbishop 

I have generated production builds to present how FloatingToolbar works

You can download it here:
iOS: **waiting to finish build**
Andorid: [APK 45971](https://github.com/wordpress-mobile/WordPress-Android/pull/11692#issuecomment-615247791)

pinarol commented:
@dratwas @cameronvoell ready for another pass.

iamthomasbishop commented:
@jbinda this is feeling pretty solid overall, other than some minor scroll position issues (mentioned below). At this point, I think I'm comfortable shipping with the toolbar (temporarily) positioned differently between platforms (iOS bottom, Android top) until we have some time with each version and can make a decision on which to go with. We can address the challenges presented on Android if bottom-positioned wins out.

## Blockers

- There are some small inconsistencies with scrolling position when selecting blocks but I think those aren't specific to the toolbar. Android especially doesn't seem to handle scroll position quite as smoothly. If these are quick refinements and related to this work, let's solve now — otherwise don't consider a blocker.
- [iOS] One thing I think we probably should do, however, is add a small bit of additional spacing between the selected block and the toolbar when the block is nearby. By this, I mean adjust the y-position when a new block is selected so that there is at least ~8px spacing between the bottom-positioned floatingToolbar and the bottom of the selected block. ([Example](https://cloudup.com/c1_davZxDjH))
- Not necessarily a "blocker" per sé, but I would like to adjust the `background-color` of the floating toolbar to be the same color as the block toolbar. If my color picker in Figma is accurate, that looks like `#2e2e2e` — if there's a variable that represents that, let's use that to be consistent.

## Non-blockers

- Can we scroll the canvas to the associated block when the selected block part of the breadcrumbs is tapped?
- It'd be nice to apply some subtle transitions, if possible:
    - Upon entry: `300ms` transition using `easeOut`, which would fade in and translate the y-position ~4-8px upward
    - Upon entry: quick `50ms` transition using `easeIn`, which would fade out and translate the y-position ~4-8px downward
    - [Here's](https://cloudup.com/c5zqWU5j0J6) what the animation would look like together

## Other

- Maybe unrelated, but when a new block is inserted on to the canvas, the scroll position does not adjust ([example](https://cloudup.com/ciDKoEgtkIP)). This happens on both platforms.

---

Great work on this! While this wasn't the original "ideal" toolbar, it is certainly an improvement to what currently exists. Thank you!

---

> Just for sake of clarity, does this mean until we come up with a new design that would take up the full-width in an area at top of the keyboard ?

@pinarol No, I meant if we can solve in a way that allows us to place the toolbar on the bottom edge, like iOS. Although as I mentioned above, I'm comfortable with differing positions for now, because it's a real improvement over what we currently have.

pinarol commented:
> @pinarol No, I meant if we can solve in a way that allows us to place the toolbar on the bottom edge, like iOS. Although as I mentioned above, I'm comfortable with differing positions for now, because it's a real improvement over what we currently have.

Actually the placement at the bottom isn’t the problem but covering the caret is. I don’t even know if there’s a way on Android to customize the scroll amount relative to caret to prevent the floating toolbar from covering the caret. I wouldn’t bet on it to happen soon considering the amount of work it took on iOS. But i am no Android expert so I’d like to get some opinions from @hypest @maxme. One thing i know for sure is, if we make the floating toolbar take up the full-width on top of the current toolbar then there will be no problem.

iamthomasbishop commented:
> Actually the placement at the bottom isn’t the problem but covering the caret is.

Yea, I just mean the general positioning of the UI component (bottom vs. top) — if/when we were able to address that on Android — but it sounds like that's not likely. So I am okay with shipping Android top-aligned, then we can iterate if it causes problems. I don't think top-aligned is necessarily a terrible option, just less discoverable and less reachable than bottom-aligned.

> One thing i know for sure is, if we make the floating toolbar take up the full-width on top of the current toolbar then there will be no problem.

I would still prefer a top-aligned truly "floating" toolbar over a docked full-width toolbar, but if this performs poorly I do have some ideas for how a full-width toolbar would look.

jbinda commented:
## Blockers

> - There are some small inconsistencies with scrolling position when selecting blocks but I think those aren't specific to the toolbar. Android especially doesn't seem to handle scroll position quite as smoothly. If these are quick refinements and related to this work, let's solve now — otherwise don't consider a blocker.


> - [iOS] One thing I think we probably should do, however, is add a small bit of additional spacing between the selected block and the toolbar when the block is nearby. By this, I mean adjust the y-position when a new block is selected so that there is at least ~8px spacing between the bottom-positioned floatingToolbar and the bottom of the selected block. ([Example](https://cloudup.com/c1_davZxDjH))

Fixed!
<img src="https://user-images.githubusercontent.com/21242757/79724042-294a1a80-82e7-11ea-9221-bce69ee9fc0f.png" width="300" />


> - Not necessarily a "blocker" per sé, but I would like to adjust the `background-color` of the floating toolbar to be the same color as the block toolbar. If my color picker in Figma is accurate, that looks like `#2e2e2e` — if there's a variable that represents that, let's use that to be consistent.

`gray-500` was used. I switched to `$background-dark-elevated` (equivalent of `#2e2e2e`)
 
See below screen for light and dark mode

<img width="250" src="https://user-images.githubusercontent.com/21242757/79725921-3caab500-82ea-11ea-8168-8d1c9ed8d3cc.png"><img width="250" src="https://user-images.githubusercontent.com/21242757/79725948-492f0d80-82ea-11ea-9d5b-9d57e2aeec9e.png">

## Non-blockers

>- Can we scroll the canvas to the associated block when the selected block part of the breadcrumbs is tapped?

I have started to investigate how to implement that according to our convo on Slack but do not have worked version because of focusing on behaviour with text blocks.

>- It'd be nice to apply some subtle transitions, if possible:
    - Upon entry: `300ms` transition using `easeOut`, which would fade in and translate the y-position ~4-8px upward
    - Upon entry: quick `50ms` transition using `easeIn`, which would fade out and translate the y-position ~4-8px downward
    - [Here's](https://cloudup.com/c5zqWU5j0J6) what the animation would look like together

It's doable and I think it's not a big deal. Do we want to include that feature in this PR @pinarol ?

## Others

>Maybe unrelated, but when a new block is inserted on to the canvas, the scroll position does not adjust (example). This happens on both platforms.

I don't think it's related to FloatingToolbar itself. It's rather something with autoscroll

jbinda committed:
adjust style of FLoatingToolbar

pinarol commented:
> > Can we scroll the canvas to the associated block when the selected block part of the breadcrumbs is tapped?
> I have started to investigate how to implement that according to our convo on Slack but do not have worked version because of focusing on behaviour with text blocks.

This was tried as a part of [this PR](https://github.com/WordPress/gutenberg/pull/18791) in a slightly different way(the aim was to scroll to the newly created block but the mechanism needed was the same) and it was reverted because of multiple problems. This is not a straightforward thing to do, so I suggest let's not block this PR for that. Buttons block would really benefit from having this version of the FloatingToolbar because the current one pushes the button down and it looks jumpy. Same for Columns when stacked horizontally.

pinarol commented:
> > It'd be nice to apply some subtle transitions, if possible:
> > - Upon entry: 300ms transition using easeOut, which would fade in and translate the y-position ~4-8px upward
> > - Upon entry: quick 50ms transition using easeIn, which would fade out and translate the y-position ~4-8px downward
> > - Here's what the animation would look like together
>
> It's doable and I think it's not a big deal. Do we want to include that feature in this PR @pinarol ?
> 

I suggest starting implementing this right after this PR considering that this was listed below non-blockers. I really prefer to have this PR merged asap to get rid of the jumpiness of individual Button blocks when stacked horizontally with other Buttons.

jbinda commented:
>I suggest starting implementing this right after this PR considering that this was listed below non-blockers. I really prefer to have this PR merged asap to get rid of the jumpiness of individual Button blocks when stacked horizontally with other Buttons.

Ok so I will prepare builds after adjustments pointed as a blockers. After it gets Thomas approve as well as code review approve from @dratwas it can be merge

dratwas commented on packages/block-editor/src/components/block-list/subdirectory-icon.js:
```diff
@@ -3,15 +3,18 @@
-		<Path d="M19 15l-6 6-1.42-1.42L15.17 16H4V4h2v10h9.17l-3.59-3.58L13 9l6 6z" />
+		<Path
+			d="M19 15l-6 6-1.42-1.42L15.17 16H4V4h2v10h9.17l-3.59-3.58L13 9l6 6z"
+			transform={ `${ isRTL ? 'scale(-1,1) translate(-20,0)' : '' }` }
```
I'm wondering if some actions are taken when we pass an empty string to the `transfrom` and if it wouldn't be better to pass `undefined` instead. I'm just thinking aloud :)

dratwas commented on packages/block-editor/src/components/floating-toolbar/index.native.js:
```diff
@@ -0,0 +1,81 @@
+/**
```
It's not a blocker but I think this file should be moved instead of copied. In the current approach, we lose history.

dratwas commented on packages/block-editor/src/components/floating-toolbar/index.native.js:
```diff
@@ -0,0 +1,81 @@
+/**
+ * External dependencies
+ */
+import React from 'react';
```
I believe it is added automatically and can be removed :)

jbinda replied:
removed

dratwas commented on packages/block-editor/src/components/floating-toolbar/index.native.js:
```diff
@@ -0,0 +1,81 @@
+			getBlockCount,
+		} = select( 'core/block-editor' );
+
+		const clientId = getSelectedBlockClientId();
```
I think it should be `selectedClientId` since the `clientId` is a bit confusing

jbinda replied:
changed with below proposal

dratwas commented on packages/block-editor/src/components/floating-toolbar/index.native.js:
```diff
@@ -0,0 +1,81 @@
+
+		const clientId = getSelectedBlockClientId();
+
+		const isSelected = !! clientId;
```
This could be removed after renaming the `clientId`

jbinda replied:
done

dratwas commented on packages/block-editor/src/components/floating-toolbar/index.native.js:
```diff
@@ -0,0 +1,81 @@
+		const clientId = getSelectedBlockClientId();
+
+		const isSelected = !! clientId;
+		const rootBlockId = getBlockHierarchyRootClientId( clientId );
+		const parentId = getBlockRootClientId( clientId );
+		const hasRootInnerBlocks = !! getBlockCount( rootBlockId );
+
+		const showFloatingToolbar = isSelected && hasRootInnerBlocks;
+
+		return {
+			clientId,
+			showFloatingToolbar,
+			parentId,
+		};
```
I think it would be better to add one check before all selectors since all of them depends on `selectedClientId`

Something like 
```js
if( !! selectedClientId ) {
  const rootBlockId = getBlockHierarchyRootClientId( selectedClientId );
  const parentId = getBlockRootClientId( selectedClientId );
  const hasRootInnerBlocks = !! getBlockCount( rootBlockId );
  return {
    selectedClientId,
    showFloatingToolbar,
    parentId,
  }
}
return { }
```

jbinda replied:
how about that:

```jsx
		const selectedClientId = getSelectedBlockClientId();

		if( ! selectedClientId ) return;

		const rootBlockId = getBlockHierarchyRootClientId( selectedClientId );

		return {
			selectedClientId,
			showFloatingToolbar: !! getBlockCount( rootBlockId ),
			parentId: getBlockRootClientId( selectedClientId ),
		};
```

jbinda replied:
WDYT ?

dratwas replied:
LGTM :)

jbinda committed:
code review changes

jbinda committed:
Merge remote-tracking branch 'origin/master' into rnmobile/feat/fixed-floating-toolbar-position

jbinda commented:
Refreshed build are here:
iOS: [still building](https://github.com/wordpress-mobile/WordPress-iOS/pull/13927#issuecomment-615245960)
Andorid: [APK 46171](https://github.com/wordpress-mobile/WordPress-Android/pull/11692#issuecomment-615247791)

@iamthomasbishop it also includes merge Button so you can check how FloatingToolbar works with them as well

iamthomasbishop commented:
@jbinda This is feeling pretty solid! The visual changes look good, and I don't think any of the additional requests I had should stop this from iteration shipping (the animations, scrolling to the selected block, etc). Great work!!

iamthomasbishop commented:
@jbinda One tiny thing: I noticed while testing on Android that the toolbar background isn't using the updated `$background-dark-elevated` color. I tried un-installing and re-installing to make sure I wasn't on the previous build, but I'm on the updated one afaik.

jbinda commented:
@iamthomasbishop ok I will check that and fix

iamthomasbishop commented:
Oh, that's odd. Maybe I have the wrong build on Android 🤔 I am seeing this on Android:

| **Light** | **Dark** |
| --------------- | --------------- |
| <a href="https://user-images.githubusercontent.com/1200432/79870185-37ac3b00-83a8-11ea-954c-131ba9c5234f.png"><img src="https://user-images.githubusercontent.com/1200432/79870185-37ac3b00-83a8-11ea-954c-131ba9c5234f.png" width="360"></a> | <a href="https://user-images.githubusercontent.com/1200432/79870257-54487300-83a8-11ea-89d2-3368399982e3.png"><img src="https://user-images.githubusercontent.com/1200432/79870257-54487300-83a8-11ea-89d2-3368399982e3.png" width="360"></a> |

jbinda commented:
As discussed on Slack I have also spot that issue on different Emulator with Android 9 (which does not support dar theme) and be able to fix that so probably will prepare fresh build (only for Android) to confirm on your side.

jbinda committed:
fix backgroundColor on Android

cameronvoell commented on packages/edit-post/src/components/layout/index.native.js:
```diff
@@ -149,6 +153,7 @@ class Layout extends Component {
 								visible={ isTemplatePickerVisible }
 							/>
 						) }
+						{ Platform.OS === 'ios' && <FloatingToolbar /> }
```
Cleaner here in `layout` than in the old BlockListBlock location, nice!

cameronvoell reviewed (approved):
This change is working well in my tests on iOS and Android. Top of screen location on Android seems like a workable solution to me, seeing the limitations related to keyboard handling. Great work!

:shipit:

jbinda commented:
Thank you al for checking this !
